File: Irrgarten-ruby/Irrgarten/player.rb
# frozen_string_literal: true
require_relative 'dice'
require_relative 'orientation'
require_relative 'directions'
module Irrgarten

class Player < LabyrinthCharacter
  @@MAX_WEAPONS = 2
  @@MAX_SHIELDS = 3
  @@INITIAL_HEALTH = 1
  @@HITS2LOSE = 3

  def initialize(num,i,s)
    super("Player #{num}",i,s,@@INITIAL_HEALTH)
    @number = num
    @consecutive_hits = 0
    @weapons= Array.new
    @shields = Array.new
  end

  def copy(other)
    super(other)
    #@number = other.get_number
    #@health = @@INITIAL_HEALTH
    @consecutive_hits = 0
    @weapons = other.weapons.dup
    @shields = other.shields.dup


  end

  def resurrect
    @weapons.clear
    @shields.clear
    reset_hits
    @health = @@INITIAL_HEALTH
  end

  def get_number
    @number
  end

  def move (direction,valid_moves)
    size = valid_moves.size
    contained = valid_moves.include?(direction)
    if((size > 0) && (!contained))
      return valid_moves[0]
    else
      return direction
    end
  end

  def attack
    @strength + sum_weapons
  end

  def defend(received_attack)
    manage_hit(received_attack)
  end

  def receive_reward
    wReward = Dice.weapons_reward
    sReward = Dice.shields_reward
    (0..wReward).each do
      wnew = new_weapon
      receive_weapon(wnew)
    end
    (0..sReward).each do
      snew = new_shield
      receive_weapon(snew)
    end
    extra_health = Dice.health_reward
    @health += extra_health
  end

  def weapons_to_string
    armas = ""
    @weapons.each do |weapon|
      armas += weapon.to_s + "\n"
    end
    armas
  end

  def shields_to_string
    escudos = ""
    @shields.each do |weapon|
      escudos += weapon.to_s + "\n"
    end
    escudos
  end


  def to_string
    "P[#{super.to_s},#{weapons_to_string},#{shields_to_string},#{@consecutive_hits}]"   #OJO COMPROBAR
  end

  private

  def receive_weapon(w)
    @weapons.reverse_each do |e|
      discard = e.discard
      if(discard)
        @weapons.delete(e)
      end
    end
    size = @weapons.length
    if(size<@@MAX_WEAPONS)
      @weapons.push(w)
    end
  end

  def receive_shield(s)
    @shields.reverse_each do |e|
      discard =  e.discard
      if(discard)
        @shields.delete(e)
      end
    end
    size = @shields.length
    if(size<@@MAX_SHIELDS)
      @shields.push(s)
    end
  end

  def new_weapon
    Weapon.new(Dice.weapon_power, Dice.uses_left)
  end

  def new_shield
    Shield.new(Dice.shield_power, Dice.uses_left)
  end

  def manage_hit(received_attack)
    defense = defensive_energy
    if(defense<received_attack)
      got_wounded
      inc_consecutive_hits
    else
      reset_hits
    end

    if((@consecutive_hits == @@HITS2LOSE)||(dead))
      reset_hits
      lose = true
    else
      lose = false
    end
    lose
  end

  def reset_hits
    @consecutive_hits = 0
  end

  def inc_consecutive_hits
    @consecutive_hits +=1
  end
  
  protected
  def sum_shields
    suma = 0
    @shields.each do |s|
      suma += s.protect
    end
    suma
  end

  def sum_weapons
    suma = 0
    @weapons.each do |s|
      suma +=  s.attack
    end
    suma
  end

  def defensive_energy
    @intelligence + sum_shields()
  end

  public_class_method :new
end
end

File: Irrgarten-Java/src/irrgarten/Player.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */

package irrgarten;
import java.util.ArrayList;
/**
 *
 * @author jhony
 */
public class Player extends LabyrinthCharacter {
    private static int MAX_WEAPONS = 2;
    private static int MAX_SHIELDS = 3;
    private static int INITIAL_HEALTH = 10;
    private static int HITS2LOSE = 3;
   
    private char number;
    
    private int consecutiveHits; 
    private ArrayList <Weapon> weapons; 
    private ArrayList <Shield> shields;
    private WeaponCardDeck weaponCardDeck;
    private ShieldCardDeck shieldCardDeck;
            
    
    Player (char number, float intelligence, float strength){                                    
        super("Player#" + number, intelligence, strength, INITIAL_HEALTH);
        shields = new ArrayList<>();
        weapons = new ArrayList<>();
        weaponCardDeck = new WeaponCardDeck();
        shieldCardDeck = new ShieldCardDeck();
        consecutiveHits = 0;
        this.number = number;
    }
    
    Player (Player other){
        super(other);
        //number = other.number;
        consecutiveHits = other.consecutiveHits;
        weapons = other.weapons;
        shields = other.shields;
        weaponCardDeck = other.weaponCardDeck;
        shieldCardDeck = other.shieldCardDeck;
    }
    
    public void resurrect(){                      
        setHealth(INITIAL_HEALTH);
        consecutiveHits = 0;
        weapons.clear();
        shields.clear();
    }
  
    public char getNumber(){
        return number;
    }
    
    public Directions move(Directions direction,ArrayList<Directions> validMoves){                 
        int size = validMoves.size();
        boolean contained = validMoves.contains(direction);
        if(size>0 && !contained)
            return validMoves.get(0);                                       
        else return direction; 
    }
    
    public float attack(){                         
        return getStrength() + sumWeapons(); 
    }
    
    public boolean defend(float receivedAttack){
        return manageHit(receivedAttack);
    }
    
    public void receiveReward(){                                                                       
        int wReward = Dice.weaponsReward();
        int sReward = Dice.shieldsReward();
        
        for (int i=0;i<wReward;i++){
            receiveWeapon(weaponCardDeck.nextCard());
        }
        
        for (int i=0;i<sReward;i++){
            receiveShield(shieldCardDeck.nextCard());
        }

        int extraHealth = Dice.healthReward();
        setHealth(getHealth() + extraHealth);
        
    }
    
    public String weaponsToString(){
        String armas="";
        for(int i=0; i<weapons.size();i++){
            armas += weapons.get(i).toString();
            armas += "\n";
        }
        return armas;
       
    }
    
    public String shieldsToString(){
        String escudos="";
        for(int i=0; i<shields.size();i++){
            escudos += shields.get(i).toString();
            escudos += "\n";
        }
        return escudos;
    }
        
    public String toString(){
        return "P["+super.toString()+","+weaponsToString()+","+shieldsToString()+","+consecutiveHits+"]";
    }
    
    private void receiveWeapon(Weapon w){
        for(int i=weapons.size()-1;i>=0;i--){
            boolean discard = weapons.get(i).discard();
            if(discard)
                weapons.remove(weapons.get(i));
        }
        int size = weapons.size();
        if(size < MAX_WEAPONS)
            weapons.add(w);
    }
    
    private void receiveShield(Shield s){
        for(int i=shields.size()-1;i>=0;i--){
            boolean discard = shields.get(i).discard();
            if(discard)
                shields.remove(shields.get(i));
        }
        int size = shields.size();
        if(size < MAX_SHIELDS)
            shields.add(s);
    }
    
    private Weapon newWeapon(){                                                     
        return new Weapon(Dice.weaponPower(), Dice.usesLeft());
    }
    
    private Shield newShield(){                                                    
        return new Shield(Dice.shieldPower(), Dice.usesLeft());
    }
    
    protected float sumWeapons(){                                                     
        float suma = 0;
        for (int i=0; i<weapons.size(); i++){
            suma+= weapons.get(i).attack(); 
        }                               
        return suma; 
    }
    
    protected float sumShields(){                                                      
        float suma = 0;
        for (int i=0; i<shields.size(); i++){
            suma+=shields.get(i).protect();
        }
        return suma;
    }
    protected float defensiveEnergy(){
        return getIntelligence() + sumShields(); 
    }
    
    private boolean manageHit(float receivedAttack){                                
        boolean lose; 
        float defense = defensiveEnergy();
        if(defense<receivedAttack){
            gotWounded();
            incConsecutiveHits();
        }else resetHits(); 
        
        if((consecutiveHits == HITS2LOSE) || (dead())){
            resetHits();
            lose = true; 
        }else lose = false;
        
        return lose; 
    }
    
    private void resetHits(){                                                       
        consecutiveHits = 0;
    }
    
    private void incConsecutiveHits(){                                              
        consecutiveHits++;
    }
    
    
    
    
    
   
    
    
    
    
}